apollo-router/src/plugins/telemetry/metrics/apollo/studio.rs

```diff
@@ -123,30 +90,58 @@ pub(crate) struct FieldStat {
 }
 
 #[derive(Default, Serialize)]
-pub(crate) struct AggregatedMetrics {
-    key: MetricsKey,
+pub(crate) struct AggregatedReport {
+    traces_per_query: HashMap<String, AggregatedTracesAndStats>,
+    operation_count: u64,
+}
+
+impl AddAssign<Report> for AggregatedReport {
+    fn add_assign(&mut self, report: Report) {
+        for (k, v) in report.traces_and_stats {
+            *self.traces_per_query.entry(k).or_default() += v;
+        }
+
+        self.operation_count += report.operation_count;
+    }
+}
+
+#[derive(Default, Debug, Serialize)]
+pub(crate) struct AggregatedTracesAndStats {
+    #[serde(with = "vectorize")]
+    pub(crate) stats_with_context: HashMap<StatsContext, AggregatedContextualizedStats>,
+    pub(crate) referenced_fields_by_type: HashMap<String, ReferencedFieldsForType>,
+}
+
+impl AddAssign<TracesAndStats> for AggregatedTracesAndStats {
+    fn add_assign(&mut self, stats: TracesAndStats) {
+        *self
+            .stats_with_context
+            .entry(stats.stats_with_context.context.clone())
+            .or_default() += stats.stats_with_context;
+
+        // No merging required here because references fields by type will always be the same for each stats report key.
+        self.referenced_fields_by_type = stats.referenced_fields_by_type;
+    }
+}
+
+#[derive(Default, Debug, Serialize)]
+pub(crate) struct AggregatedContextualizedStats {
+    context: StatsContext,
```

Comment:
To use only one set of types we would need to put the aggregated report in a mutex and swap it out on each send. This is an option that we could have used, but maybe it is safer not to do that in case of contention, but we can revisit another time.

Let's do any renaming at the end once we are sure everything is working.

Reply:
> Doesn't this struct live in a `HashMap<StatsContext, AggregatedContextualizedStats>` already? Seems like this field is redundant? Note that in protobufs, map keys can only be strings or integers, which is we we have a `repeated ContextualizedStats` with a context field on it instead of a map. But you're already using a map here (since rust apparently supports more general map keys) so this seems like it should be redundant?
> 
> I guess you'd need to change the definition of
> 
> ```
> impl From<AggregatedTracesAndStats> for apollo_spaceport::TracesAndStats {
> ```
> 
> to create stats_with_context from map entries rather than values.

This won't work if I want to use: `impl From<AggregatedContextualizedStats> for apollo_spaceport::ContextualizedStats`
as From does not have access to the entry that it was stored in.

For consistency with the rest of the conversion logic lets keep it as is for now and look again if it is a perf problem.

Reply:
Happy to trust you on this one! Maybe add a comment noting the redundancy if you think that's helpful.

Reply:
Re

> To use only one set of types we would need to put the aggregated report in a mutex and swap it out on each send. This is an option that we could have used, but maybe it is safer not to do that in case of contention. We can revisit another time.

I don't think I understand this one — it seems like the AggregatedReport type can be used anywhere the Report type can be used, *except* if you want to be able to put a duration into it and pull it out again without rounding. But my feedback is merely about clarity/simplicity/code size, and I very well may be missing something important. No need to convince me of this, happy to trust you're making the right call for today and quite possibly for tomorrow as well.

Reply:
What I'd really like is that the generated rs files for the protobuf somehow can use the Histogram type. Didn't get time to investigate this though.

I think in the long run we'll make the changes to the datastructure to convert many of the u64s to bools as you suggested earlier, so maybe things will diverge in the long run.